src/main/java/seedu/duke/storage/FoodListDecoder.java

```diff
@@ -1,49 +1,233 @@
+     * @return a Food object with attributes
+     * @throws ReadStorageFileErrorException when there are errors with decoding the attributes
+     * @throws EmptyStorageFileException  when the line is empty
      */
-    private static Food decodeFoodFromString(String encodedFood) throws DukeException {
-        String[] details = encodedFood.split("«");
+    private static Food decodeFoodFromString(String encodedFood) throws
+            ReadStorageFileErrorException, EmptyStorageFileException {
 
+        if (encodedFood.equals("")) {
+            throw new EmptyStorageFileException();
+        }
+        String[] details = encodedFood.split(SEPARATOR);
+        String foodName = "";
+        String expiryDate = "";
+        String quantity = "0.0";
+        String unit = "";
+        String category = "";
+
+
+        for (String foodDetails : details) {
+            try {
+                if (foodDetails.equals("")) {
+                    continue;
+                }
+
+                char identifier = foodDetails.charAt(0);
+                String detail = foodDetails.substring(1).trim();
+
+                switch(identifier) {
+                case NAME_IDENTIFIER:
+                    if (hasValidFoodName(detail)) {
+                        assert !detail.equals("");
+                        foodName = detail;
+                    }
+                    break;
+                case EXPIRY_DATE_IDENTIFIER:
+                    if (hasValidExpiryDate(detail)) {
+                        expiryDate = detail;
+                    }
+                    break;
+                case QUANTITY_IDENTIFIER:
+                    if(hasValidQuantity(detail)) {
+                        quantity = detail;
+                    }
+                    break;
+                case UNIT_IDENTIFIER:
+                    if (hasValidUnit(detail)) {
+                        unit = detail;
+                    } else {
+                        unit = String.valueOf(Unit.UNIT);
+                    }
+                    break;
+                case CATEGORY_IDENTIFIER:
+                    if(hasValidCategory(detail)) {
+                        category = detail;
+                    } else {
+                        category = String.valueOf(FoodCategory.UNCLASSIFIED_FOOD);
+                    }
+                    break;
+                default:
+                    break;
+                }
+            } catch (StringIndexOutOfBoundsException e) {
+                Ui.showError(MISSING_IDENTIFIER_ERROR);
+            }
+        }
         try {
-            return new Food(details[0].trim(), details[1].trim()
-                    , parseDouble(details[2].trim()), details[3].trim(), details[4].trim());
-        } catch (ArrayIndexOutOfBoundsException e) {
-            throw new StorageOperationException("Unable to read file successfully");
+            return new Food(foodName, expiryDate, Double.valueOf(quantity), unit, category);

```

Comment:
the previous return statement is very confusing and now you have changed the way of representing each details which is a good improvement in the code quality.